File: servers/login/src/test/kotlin/fr/rob/login/test/feature/service/network/Listener.kt
package fr.rob.login.test.feature.service.network

import fr.rob.core.misc.clock.IntervalTimer
import fr.rob.core.network.Packet

class Listener(private val callback: (opcode: Int, packet: Packet, msg: Any?) -> Boolean, timeout: Int) {

    private val timer = IntervalTimer(timeout)

    var isOutdated: Boolean = false
    var isDone: Boolean = false

    fun call(opcode: Int, packet: Packet, any: Any?) {
        isDone = callback(opcode, packet, any)
    }

    fun update(delta: Int) {
        if (isDone || isOutdated) {
            return
        }

        timer.update(delta)

        if (timer.passed()) {
            isOutdated = true
        }
    }
}

File: RobClient/src/Listener.cs
using Fr.Raven.Proto.Message.Gateway;

namespace RobClient;

class Listener {
    private Func<Packet, Boolean> _callback;

    public Listener(Func<Packet, Boolean> callback) {
        _callback = callback;
    }

    public Boolean Call(Packet packet) {
        return _callback(packet);
    }
}